File: src/main/java/com/meti/api/magma/collect/ArrayList.java
package com.meti.api.magma.collect;

public record ArrayList<T>(Object[] elements, int elementSize) implements List<T> {
	@Override
	public List<T> add(T element) throws CollectionException {
		try {
			return set(elementSize, element);
		} catch (IndexException e) {
			throw new CollectionException(e);
		}
	}

	@Override
	public List<T> set(int index, T element) throws IndexException {
		if (index < 0) {
			var format = "Index of '%d' is negative.";
			var message = format.formatted(index);
			throw new IndexException(message);
		}

		var newElements = index < elements.length ? elements : resize(index);
		var newSize = index + 1 > elementSize ? index + 1 : elementSize;
		newElements[index] = element;
		return new ArrayList<>(newElements, newSize);
	}

	public Object[] resize(int index) {
		var newCapacity = calculateCapacity(index);
		var copy = new Object[newCapacity];
		for (int i = 0; i < elementSize; i++) copy[i] = elements[i];
		return copy;
	}

	private int calculateCapacity(int index) {
		var oldCapacity = elements.length;
		var current = oldCapacity == 0 ? 1 : oldCapacity;
		while (current <= index) {
			current *= 2;
		}
		return current;
	}

	@Override
	public T apply(int index) throws IndexException {
		if (index < 0) {
			var format = "Index of '%d' is negative.";
			var message = format.formatted(index);
			throw new IndexException(message);
		}
		if (index >= elementSize) {
			var format = "Index of '%d' is greater than or equal to length.";
			var message = format.formatted(index);
			throw new IndexException(message);
		}
		return (T) elements[index];
	}

	@Override
	public int size() {
		return elementSize;
	}
}

File: src/main/java/com/meti/ArrayList.scala
package com.meti

object ArrayList {
  def empty[T] = new ArrayList[T](new Array[Any](10), 0)
}

class ArrayList[T](val elements: Array[Any], val elementsSize: Int) extends List[T] {
  @throws[CollectionException]
  override def add(element: T): List[T] = set(elementsSize, element)

  @throws[CollectionException]
  override def set(index: Int, element: T): List[T] = {
    if (index < 0) throw new IndexException("Index is negative.")
    val newElements = if (index < elements.length) elements else resize(index)
    val newSize = if (index + 1 > elementsSize) index + 1 else elementsSize
    newElements(index) = element
    new ArrayList[T](newElements, newSize)
  }

  private def resize(index: Int): Array[Any] = {
    val copy = new Array[Any](calculateCapacity(index))
    for (i <- 0 until elementsSize) {
      copy(i) = elements(i)
    }
    copy
  }

  private def calculateCapacity(index: Int) = {
    var toUse = if (elements.length == 0) 1
    else elements.length
    while (index >= toUse) toUse *= 2
    toUse
  }

  override def stream: Stream[T] = new AbstractStream[T]() {
    private var counter = 0

    @throws[StreamException]
    override def head: T = if (counter < elementsSize) {
      counter += 1
      elements(counter - 1).asInstanceOf[T]
    }
    else throw EndOfStreamException("No more elements in list left.")
  }
}